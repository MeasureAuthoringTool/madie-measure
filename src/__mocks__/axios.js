const mock = jest.genMockFromModule("axios");
const axios = jest.requireActual("axios");

jest.mock("axios", () => {
  return {
    create: jest.fn(() => ({
      get: jest.fn(),
      interceptors: {
        request: { use: jest.fn(), eject: jest.fn() },
        response: { use: jest.fn(), eject: jest.fn() },
      },
    })),
    get: jest.fn(),
  };
});

module.exports = axios;
