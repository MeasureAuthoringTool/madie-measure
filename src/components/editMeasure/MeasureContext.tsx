import { createContext } from "react";
import { Measure } from "../../../../madie-models/dist";

export interface MeasureContextHolder {
  measure: Measure;
  setMeasure: (measure: Measure) => void;
}

const MeasureContext = createContext<MeasureContextHolder>(null);

export default MeasureContext;

export const MeasureContextProvider = MeasureContext.Provider;
export const MeasureContextConsumer = MeasureContext.Consumer;
