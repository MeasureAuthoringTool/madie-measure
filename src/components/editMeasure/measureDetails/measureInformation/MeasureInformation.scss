.textinputnoborder {
  border: none;
  margin-top: -8px;
}

.textlabel {
  color: black;
  font-family: Rubik, sans-serif;
  font-weight: 450px;
  text-transform: capitalize;
  padding-top: -10;
}
