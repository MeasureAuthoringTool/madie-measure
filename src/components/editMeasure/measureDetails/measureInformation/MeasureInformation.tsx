import React, { useState, useEffect } from "react";
import { useHistory } from "react-router-dom";
import tw from "twin.macro";
import styled, { css } from "styled-components";
import { Measure } from "@madie/madie-models";
import useMeasureServiceApi from "../../../../api/useMeasureServiceApi";
import "styled-components/macro";
import {
  Button,
  Toast,
  TextField,
} from "@madie/madie-design-system/dist/react";
import DeleteDialog from "./DeleteDialog";
import LocalizationProvider from "@mui/lab/LocalizationProvider";
import DateAdapter from "@mui/lab/AdapterDateFns";
import DesktopDatePicker from "@mui/lab/DesktopDatePicker";
import { DialogContent, DialogTitle, Divider, Typography } from "@mui/material";
import { useFormik } from "formik";
import { HelperText } from "@madie/madie-components";
import { MeasureSchemaValidator } from "../../../../validations/MeasureSchemaValidator";
import { measureStore, useOktaTokens } from "@madie/madie-util";
import classNames from "classnames";
import { makeStyles } from "@mui/styles";
import { Box } from "@mui/system";
import { synchingEditorCqlContent } from "@madie/madie-editor";

interface measureInformationForm {
  measureName: string;
  cqlLibraryName: string;
  ecqmTitle: string;
  measurementPeriodStart: Date;
  measurementPeriodEnd: Date;
}

const Form = tw.form`max-w-xl my-8`;
const MessageDiv = tw.div`ml-3`;
const MessageText = tw.p`text-sm font-medium`;
const SuccessText = tw(MessageText)`text-green-800`;
const ErrorText = tw(MessageText)`text-red-800`;

export default function MeasureInformation() {
  const history = useHistory();
  const measureServiceApi = useMeasureServiceApi();
  const { updateMeasure } = measureStore;
  const [measure, setMeasure] = useState<any>(measureStore.state);
  useEffect(() => {
    const subscription = measureStore.subscribe(setMeasure);
    return () => {
      subscription.unsubscribe();
    };
  }, []);
  const [genericErrorMessage, setGenericErrorMessage] = useState<string>();
  const { getUserName } = useOktaTokens();
  const userName = getUserName();

  const useStyles = makeStyles({
    row: {
      display: "flex",
      flexDirection: "row",
    },
    end: {
      justifyContent: "flex-end",
    },
    dialogTitle: {
      display: "flex",
      flexDirection: "row",
      justifyContent: "space-between",
      alignItems: "center",
      padding: "24px 32px",
    },
    title: {
      fontFamily: "Rubik",
      fontSize: 24,
      padding: 0,
    },
    info: {
      fontSize: 14,
      fontWeight: 300,
      fontFamily: "Rubik",
    },
    asterisk: {
      color: "#D92F2F",
      marginRight: 3,
    },
  });
  const classes = useStyles();
  const flexEnd = classNames(classes.row, classes.end);

  const row = {
    display: "flex",
    flexDirection: "row",
  };
  const spaced = {
    marginTop: "23px",
  };
  const gap = {
    columnGap: "24px",
    "& > * ": {
      flex: 1,
    },
  };
  const formRow = Object.assign({}, row, spaced);
  const formRowGapped = Object.assign({}, formRow, gap);

  // Dialog and toast utilities
  const [deleteOpen, setDeleteOpen] = useState<boolean>(false);
  const [toastOpen, setToastOpen] = useState<boolean>(false);
  const [toastMessage, setToastMessage] = useState<string>("");
  const [toastType, setToastType] = useState<string>("danger");
  const [successMessage, setSuccessMessage] = useState<string>(null);
  const [errorMessage, setErrorMessage] = useState<string>(null);
  // our initial values are taken from the measure we subscribe to
  const INITIAL_VALUES = {
    measurementPeriodStart: measure?.measurementPeriodStart,
    measurementPeriodEnd: measure?.measurementPeriodEnd,
    measureName: measure?.measureName,
    cqlLibraryName: measure?.cqlLibraryName,
    ecqmTitle: measure?.ecqmTitle,
  } as measureInformationForm;

  const formik = useFormik({
    initialValues: { ...INITIAL_VALUES },
    validationSchema: MeasureSchemaValidator,
    enableReinitialize: true, // formik will auto set initial variables whenever measure delivers new results
    onSubmit: async (values: measureInformationForm) =>
      await handleSubmit(values),
  });

  const canEdit = measure?.createdBy === userName;
  const onToastClose = () => {
    setToastType(null);
    setToastMessage("");
    setToastOpen(false);
  };
  const handleToast = (type, message, open) => {
    setToastType(type);
    setToastMessage(message);
    setToastOpen(open);
  };

  const deleteMeasure = async () => {
    const deletedMeasure: Measure = { ...measure, active: false };
    try {
      const result = await measureServiceApi.updateMeasure(deletedMeasure);
      if (result.status === 200) {
        handleToast("success", "Measure successfully deleted", true);
        setTimeout(() => {
          history.push("/measures");
        }, 3000);
      }
    } catch (e) {
      if (e?.response?.data) {
        const { error, status, message } = e.response.data;
        const errorMessage = `${status}: ${error} ${message}`;
        handleToast("danger", errorMessage, true);
      } else {
        handleToast("danger", e.toString(), true);
      }
    }
  };

  const handleSubmit = async (values) => {
    const inSyncCql = await synchingEditorCqlContent(
      "",
      measure?.cql,
      values.cqlLibraryName,
      measure?.cqlLibraryName,
      "0.0.000", //as the versioning is not implemented in measure for now we just send default value: 0.0.000
      "measureInformation"
    );
    const newMeasure: Measure = {
      ...measure,
      measureName: values.measureName,
      cqlLibraryName: values.cqlLibraryName,
      ecqmTitle: values.ecqmTitle,
      measurementPeriodStart: values.measurementPeriodStart,
      measurementPeriodEnd: values.measurementPeriodEnd,
      cql: inSyncCql,
    };
    measureServiceApi
      .updateMeasure(newMeasure)
      .then(() => {
        setSuccessMessage("Measurement Information Updated Successfully");
        // updating measure will propagate update state site wide.
        updateMeasure(newMeasure);
      })
      .catch((err) => {
        setErrorMessage(err.response.data.message);
      });
  };

  function formikErrorHandler(name: string, isError: boolean) {
    if (formik.touched[name] && formik.errors[name]) {
      return (
        <HelperText
          data-testid={`${name}-helper-text`}
          text={formik.errors[name]?.toString()}
          isError={isError}
        />
      );
    }
  }

  return (
    <div tw="px-1 pt-1" data-testid="measure-information-edit">
      {genericErrorMessage && (
        <div tw="bg-red-500 pt-4 px-4 pb-4">
          <span
            tw="text-white"
            data-testid="edit-measure-information-generic-error-text"
          >
            {genericErrorMessage}
          </span>
        </div>
      )}

      <Form
        style={{ marginTop: 0 }}
        onSubmit={formik.handleSubmit}
        data-testid="measurement-information-form"
      >
        <div className={classes.dialogTitle}>
          <DialogTitle className={classes.title}>Information</DialogTitle>
        </div>

        <DialogContent>
          <div className={flexEnd}>
            <Typography className={classes.info}>
              <span className={classes.asterisk}>*</span>
              Indicates required field
            </Typography>
          </div>

          <div>
            <Divider style={{ marginTop: 20, marginBottom: 20 }} />
          </div>
          <Box sx={formRow}>
            <TextField
              placeholder="Measure Name"
              required
              disabled={!canEdit}
              label="Measure Name"
              id="measureName"
              inputProps={{ "data-testid": "measure-name-input" }}
              helperText={formikErrorHandler("measureName", true)}
              data-testid="measure-name-text-field"
              size="small"
              error={
                formik.touched.measureName && Boolean(formik.errors.measureName)
              }
              {...formik.getFieldProps("measureName")}
            />
          </Box>

          <Box sx={formRow}>
            <TextField
              placeholder="Enter CQL Library Name"
              required
              disabled={!canEdit}
              label="Measure CQL Library Name"
              id="cqlLibraryName"
              data-testid="cql-library-name"
              inputProps={{ "data-testid": "cql-library-name-input" }}
              helperText={formikErrorHandler("cqlLibraryName", true)}
              size="small"
              error={
                formik.touched.cqlLibraryName &&
                Boolean(formik.errors.cqlLibraryName)
              }
              {...formik.getFieldProps("cqlLibraryName")}
            />
          </Box>

          <Box sx={formRowGapped}>
            <TextField
              placeholder="eCQM Name"
              required
              disabled={!canEdit}
              label="eCQM Abbreviated Title"
              id="ecqmTitle"
              data-testid="ecqm-text-field"
              inputProps={{ "data-testid": "ecqm-input" }}
              helperText={formikErrorHandler("ecqmTitle", true)}
              size="small"
              error={
                formik.touched.ecqmTitle && Boolean(formik.errors.ecqmTitle)
              }
              {...formik.getFieldProps("ecqmTitle")}
            />
          </Box>

          <Box sx={formRowGapped} data-testid="measurement-period-div">
            <LocalizationProvider dateAdapter={DateAdapter}>
              <DesktopDatePicker
                disableOpenPicker={true}
                disabled={!canEdit}
                label="Measurement Period - Start Date"
                inputFormat="MM/dd/yyyy"
                value={formik.values.measurementPeriodStart}
                onChange={(startDate) => {
                  formik.setFieldValue("measurementPeriodStart", startDate);
                }}
                renderInput={(params) => {
                  const { onChange, ...formikFieldProps } =
                    formik.getFieldProps("measurementPeriodStart");
                  return (
                    <TextField
                      {...formikFieldProps}
                      {...params}
                      required
                      data-testid="measurement-period-start"
                      helperText={formikErrorHandler(
                        "measurementPeriodStart",
                        true
                      )}
                    />
                  );
                }}
              />
            </LocalizationProvider>
            <LocalizationProvider dateAdapter={DateAdapter}>
              <DesktopDatePicker
                disableOpenPicker={true}
                disabled={!canEdit}
                label="Measurement Period - End Date"
                inputFormat="MM/dd/yyyy"
                value={formik.values.measurementPeriodEnd}
                onChange={(endDate) => {
                  formik.setFieldValue("measurementPeriodEnd", endDate);
                }}
                renderInput={(params) => {
                  const { onChange, ...formikFieldProps } =
                    formik.getFieldProps("measurementPeriodEnd");
                  return (
                    <TextField
                      {...formikFieldProps}
                      {...params}
                      required
                      data-testid="measurement-period-end"
                      helperText={formikErrorHandler(
                        "measurementPeriodEnd",
                        true
                      )}
                    />
                  );
                }}
              />
            </LocalizationProvider>
          </Box>
        </DialogContent>

        <Button
          className="qpp-c-button--cyan"
          type="submit"
          data-testid="measurement-information-save-button"
          disabled={!(formik.isValid && formik.dirty)}
          style={{ marginTop: 20 }}
        >
          Save
        </Button>
        {canEdit && (
          <Button
            variant="danger-primary"
            data-testid="delete-measure-button"
            disabled={!canEdit}
            onClick={() => setDeleteOpen(true)}
            style={{ marginTop: 20, float: "right" }}
          >
            Delete Measure
          </Button>
        )}
      </Form>

      <MessageDiv>
        {successMessage && (
          <SuccessText data-testid="measurement-information-success-message">
            {successMessage}
          </SuccessText>
        )}
        {errorMessage && (
          <ErrorText data-testid="measurement-information-error-message">
            {errorMessage}
          </ErrorText>
        )}
      </MessageDiv>

      <DeleteDialog
        open={deleteOpen}
        onClose={() => setDeleteOpen(false)}
        measureName={measure?.measureName}
        deleteMeasure={deleteMeasure}
      />
      <Toast
        toastKey="measure-information-toast"
        toastType={toastType}
        testId={
          toastType === "danger"
            ? "edit-measure-information-generic-error-text"
            : "edit-measure-information-success-text"
        }
        open={toastOpen}
        message={toastMessage}
        onClose={onToastClose}
        autoHideDuration={6000}
      />
    </div>
  );
}
