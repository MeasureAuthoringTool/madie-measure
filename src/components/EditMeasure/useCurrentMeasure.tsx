import React, { useContext } from "react";
import MeasureContext from "./MeasureContext";
import Measure from "../../models/Measure";

export default function useCurrentMeasure() {
  const measure: Measure = useContext(MeasureContext);
  return measure;
}
