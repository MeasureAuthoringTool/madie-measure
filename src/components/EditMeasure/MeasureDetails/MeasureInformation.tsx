import React from "react";
import InlineEdit from "../../InlineEdit/InlineEdit";
import Measure from "../../../models/Measure";
import useMeasureServiceApi from "../../../api/useMeasureServiceApi";
import useCurrentMeasure from "../useCurrentMeasure";

export default function MeasureInformation() {
  const measureServiceApi = useMeasureServiceApi();
  const { measure, setMeasure } = useCurrentMeasure();

  function updateMeasureTitle(text: string): void {
    if (text !== measure.measureName) {
      const newMeasure: Measure = { ...measure, measureName: text };

      measureServiceApi
        .updateMeasure(newMeasure)
        .then(() => {
          setMeasure(newMeasure);
        })
        .catch((reason) => {
          const message = `Error updating measure ${newMeasure.measureName}`;
          console.error(message);
          console.error(reason);
          throw new Error(message);
        });
    }
  }

  return (
    <div tw="px-4" data-testid="measureNameEdit">
      <span>Measure:</span>
      <InlineEdit text={measure.measureName} onSetText={updateMeasureTitle} />
    </div>
  );
}
