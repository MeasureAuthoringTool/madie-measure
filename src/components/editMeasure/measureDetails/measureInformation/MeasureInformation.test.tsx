import * as React from "react";
import {
  render,
  cleanup,
  fireEvent,
  waitFor,
  screen,
  within,
  logRoles,
} from "@testing-library/react";
import { act } from "react-dom/test-utils";
import MeasureInformation from "./MeasureInformation";
import useMeasureServiceApi, {
  MeasureServiceApi,
} from "../../../../api/useMeasureServiceApi";
import useCurrentMeasure from "../../useCurrentMeasure";
import { MeasureContextHolder } from "../../MeasureContext";
import Measure from "../../../../models/Measure";
import { MemoryRouter } from "react-router";
import userEvent from "@testing-library/user-event";
import { useOktaTokens, useKeyPress } from "@madie/madie-util";
import { describe, expect, it } from "@jest/globals";
import axios, { AxiosError, AxiosResponse } from "axios";

const mockHistoryPush = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

jest.mock("../../../../api/useMeasureServiceApi");
jest.mock("../../useCurrentMeasure");
jest.mock("@madie/madie-util", () => ({
  useOktaTokens: jest.fn(() => ({
    getUserName: jest.fn(() => "testuser@example.com"), //#nosec
  })),
  useKeyPress: jest.fn(() => false),
  measureStore: {
    updateMeasure: jest.fn((measure) => measure),
  },
  useOnClickOutside: jest.fn(() => false),
}));

const useMeasureServiceApiMock =
  useMeasureServiceApi as jest.Mock<MeasureServiceApi>;

const useCurrentMeasureMock =
  useCurrentMeasure as jest.Mock<MeasureContextHolder>;
const axiosError: AxiosError = {
  response: {
    status: 500,
    data: { status: 500, error: "bad test", message: "oh no what happened" },
  } as AxiosResponse,
  toJSON: jest.fn(),
} as unknown as AxiosError;

let serviceApiMock: MeasureServiceApi;
serviceApiMock = {
  updateMeasure: jest
    .fn()
    .mockResolvedValueOnce(undefined)
    .mockResolvedValueOnce(undefined)
    .mockResolvedValueOnce(undefined)
    .mockResolvedValueOnce({ status: 200 })
    .mockResolvedValueOnce(undefined)
    .mockRejectedValueOnce(axiosError)
    .mockRejectedValueOnce({ response: { data: { measureName: "bad" } } })
    .mockRejectedValueOnce({ response: { data: { measureName: "bad" } } }),
} as unknown as MeasureServiceApi;
useMeasureServiceApiMock.mockImplementation(() => serviceApiMock);

useOktaTokens.mockImplementation(() => ({
  getUserName: () => "testuser@example.com", // #nosec
}));

describe("MeasureInformation component", () => {
  let measure: Measure;
  let measureContextHolder: MeasureContextHolder;
  afterEach(cleanup);

  beforeEach(() => {
    measure = {
      id: "test measure",
      measureName: "the measure for testing",
      createdBy: "testuser@example.com",
    } as Measure;

    measureContextHolder = {
      measure,
      setMeasure: jest.fn(),
    };

    useCurrentMeasureMock.mockImplementation(() => measureContextHolder);
  });
  const {
    getByTestId,
    findByTestId,
    queryByText,
    queryByTestId,
    findByRole,
    getByRole,
    findByText,
  } = screen;

  it("should render the component with measure's name populated", () => {
    render(<MeasureInformation />);
    const result: HTMLElement = getByTestId("measure-name-edit");
    expect(result).toBeInTheDocument();
    const text = getByTestId("inline-view-span");
    expect(text.textContent).toBe(measure.measureName);
  });

  it("should render the component with a blank measure name", () => {
    delete measure.measureName;
    render(<MeasureInformation />);
    const result: HTMLElement = getByTestId("measure-name-edit");
    expect(result).toBeInTheDocument();
    const text = getByTestId("inline-view-span");
    expect(text.textContent).toBe("");
  });

  it("Check if the measurement period save button is present", () => {
    render(<MeasureInformation />);
    const result: HTMLElement = getByTestId("measurement-period-save-button");
    expect(result).toBeInTheDocument();
  });

  it("Check if measurement start field is present in the form", () => {
    render(<MeasureInformation />);
    const result = getByTestId("measurement-period-form");
    expect(result).toBeInTheDocument();
    const measurementPeriodStart: HTMLElement = getByTestId(
      "measurement-period-start-date"
    );
    expect(measurementPeriodStart).toBeInTheDocument();
  });

  it("Check if measurement start date field updates input as expected", async () => {
    render(<MeasureInformation />);
    const measurementPeriodStartNode = getByTestId("measurement-period-start");
    const measurementPeriodStartInput = within(
      measurementPeriodStartNode
    ).getByRole("textbox");
    userEvent.type(measurementPeriodStartInput, "12/07/2001");
    await act(async () => {
      await waitFor(() =>
        expect(measurementPeriodStartInput.value).toBe("12/07/2001")
      );
    });
  });

  it("Check if measurement end date field has expected value", async () => {
    render(<MeasureInformation />);
    const measurementPeriodEndNode = getByTestId("measurement-period-end");
    const measurementPeriodEndInput = within(
      measurementPeriodEndNode
    ).getByRole("textbox");
    userEvent.type(measurementPeriodEndInput, "12/07/2009");
    await waitFor(() =>
      expect(measurementPeriodEndInput.value).toBe("12/07/2009")
    );
  });

  it("Check if measurement period save button is disabled when measurement period start and end date have same values", async () => {
    render(<MeasureInformation />);
    const measurementPeriodStartNode = getByTestId("measurement-period-start");
    const measurementPeriodStartInput = within(
      measurementPeriodStartNode
    ).getByRole("textbox");
    userEvent.type(measurementPeriodStartInput, "12/07/2009");
    await act(async () => {
      await waitFor(() =>
        expect(measurementPeriodStartInput.value).toBe("12/07/2009")
      );
    });
    const measurementPeriodEndNode = getByTestId("measurement-period-end");
    const measurementPeriodEndInput = within(
      measurementPeriodEndNode
    ).getByRole("textbox");
    userEvent.type(measurementPeriodEndInput, "12/07/2009");
    await waitFor(() =>
      expect(measurementPeriodEndInput.value).toBe("12/07/2009")
    );

    const createBtn = getByTestId("measurement-period-save-button");
    expect(createBtn).toBeInTheDocument();
    expect(createBtn).toBeDisabled();
  });

  it("Check if measurement period save button is disabled when measurement period end date is less than start date", async () => {
    render(<MeasureInformation />);
    const measurementPeriodStartNode = getByTestId("measurement-period-start");
    const measurementPeriodStartInput = within(
      measurementPeriodStartNode
    ).getByRole("textbox");
    userEvent.type(measurementPeriodStartInput, "12/07/2009");
    await act(async () => {
      await waitFor(() =>
        expect(measurementPeriodStartInput.value).toBe("12/07/2009")
      );
    });
    const measurementPeriodEndNode = getByTestId("measurement-period-end");
    const measurementPeriodEndInput = within(
      measurementPeriodEndNode
    ).getByRole("textbox");
    userEvent.type(measurementPeriodEndInput, "12/07/2008");
    await waitFor(() =>
      expect(measurementPeriodEndInput.value).toBe("12/07/2008")
    );

    const createBtn = getByTestId("measurement-period-save-button");
    expect(createBtn).toBeInTheDocument();
    expect(createBtn).toBeDisabled();
  });

  it("Check if measurement period save button is disabled when measurement period end date or state date is not valid", async () => {
    render(<MeasureInformation />);
    const measurementPeriodStartNode = getByTestId("measurement-period-start");
    const measurementPeriodStartInput = within(
      measurementPeriodStartNode
    ).getByRole("textbox");
    userEvent.type(measurementPeriodStartInput, "12/07/200");
    await act(async () => {
      await waitFor(() =>
        expect(measurementPeriodStartInput.value).toBe("12/07/200")
      );
    });
    const measurementPeriodEndNode = getByTestId("measurement-period-end");
    const measurementPeriodEndInput = within(
      measurementPeriodEndNode
    ).getByRole("textbox");
    userEvent.type(measurementPeriodEndInput, "12/07/2008");
    await waitFor(() =>
      expect(measurementPeriodEndInput.value).toBe("12/07/2008")
    );

    const createBtn = getByTestId("measurement-period-save-button");
    expect(createBtn).toBeInTheDocument();
    expect(createBtn).toBeDisabled();
  });

  it("Check if measurement period save button is enabled when measurement period start and end dates pass all date checks", async () => {
    render(<MeasureInformation />);
    const measurementPeriodStartNode = getByTestId("measurement-period-start");
    const measurementPeriodStartInput = within(
      measurementPeriodStartNode
    ).getByRole("textbox");
    userEvent.type(measurementPeriodStartInput, "12/07/2000");
    await act(async () => {
      await waitFor(() =>
        expect(measurementPeriodStartInput.value).toBe("12/07/2000")
      );
    });
    const measurementPeriodEndNode = getByTestId("measurement-period-end");
    const measurementPeriodEndInput = within(
      measurementPeriodEndNode
    ).getByRole("textbox");
    userEvent.type(measurementPeriodEndInput, "12/07/2009");
    await waitFor(() =>
      expect(measurementPeriodEndInput.value).toBe("12/07/2009")
    );

    const createBtn = getByTestId("measurement-period-save-button");
    expect(createBtn).toBeInTheDocument();
    expect(createBtn).toBeEnabled();
  });

  it("saving measurement periods", async () => {
    render(<MeasureInformation />);
    const measurementPeriodStartNode = getByTestId("measurement-period-start");
    const measurementPeriodStartInput = within(
      measurementPeriodStartNode
    ).getByRole("textbox");
    userEvent.type(measurementPeriodStartInput, "12/07/2000");
    await act(async () => {
      await waitFor(() =>
        expect(measurementPeriodStartInput.value).toBe("12/07/2000")
      );
    });
    const measurementPeriodEndNode = getByTestId("measurement-period-end");
    const measurementPeriodEndInput = within(
      measurementPeriodEndNode
    ).getByRole("textbox");
    userEvent.type(measurementPeriodEndInput, "12/07/2009");
    await waitFor(() =>
      expect(measurementPeriodEndInput.value).toBe("12/07/2009")
    );

    const createBtn = getByTestId("measurement-period-save-button");
    fireEvent.click(createBtn);
    await waitFor(
      () =>
        expect(
          getByTestId("measurement-period-success-message")
        ).toBeInTheDocument(),
      {
        timeout: 5000,
      }
    );
  });
  it("should save the measure's name on an update", async () => {
    render(<MeasureInformation />);
    // Click the name to trigger the inline edit
    const clickableSpan = getByTestId("inline-view-span");
    fireEvent.click(clickableSpan);
    // Have the user click enter (after input)
    useKeyPress.mockReturnValueOnce(true);
    // Type in the new value
    const input = await findByTestId("inline-edit-input");
    fireEvent.change(input, {
      target: { value: "new value" },
    });
    // Wait for rendering
    await findByTestId("inline-view-span");
    // Check expectations
    const expected = {
      id: "test measure",
      measureName: "new value",
      createdBy: "testuser@example.com",
    };
    const mockSetMeasure = measureContextHolder.setMeasure as jest.Mock<void>;
    expect(serviceApiMock.updateMeasure).toHaveBeenCalledWith(expected);
    expect(mockSetMeasure).toHaveBeenCalledWith(expected);
  });

  it("Should display a delete button if user is the owner of measure", async () => {
    render(<MeasureInformation />);
    const result: HTMLElement = await findByTestId("delete-measure-button");
    expect(result).toBeInTheDocument();
  });

  it("Should not display a delete button if user is not the owner of measure", () => {
    useOktaTokens.mockImplementationOnce(() => ({
      getUserName: () => "othertestuser@example.com", //#nosec
    }));
    render(<MeasureInformation />);
    const result: HTMLElement = queryByText("delete-measure-button");
    expect(result).toBeNull();
  });

  it("On delete click, user is presented with a confirm deletion screen", async () => {
    render(<MeasureInformation />);
    const result: HTMLElement = await findByTestId("delete-measure-button");
    fireEvent.click(result);
    const confirmDelete = getByTestId("delete-measure-button-2");
    expect(confirmDelete).toBeInTheDocument();
    const cancelDelete = await getByTestId("cancel-delete-measure-button");
    expect(cancelDelete).toBeInTheDocument();
  });

  it("On successful delete action click, user can see success message and routes back to measures", async () => {
    await render(
      <MemoryRouter>
        <MeasureInformation />
      </MemoryRouter>
    );
    const result: HTMLElement = await findByTestId("delete-measure-button");
    fireEvent.click(result);
    const confirmDelete = await getByTestId("delete-measure-button-2");
    expect(confirmDelete).toBeInTheDocument();
    fireEvent.click(confirmDelete);
    await waitFor(
      () => {
        expect(
          getByTestId("edit-measure-information-success-text")
        ).toBeInTheDocument();
        expect(mockHistoryPush).toHaveBeenCalledWith("/measures");
      },
      {
        timeout: 5000,
      }
    );
  });

  it("On failed delete action click, user can see toast error pop up", async () => {
    render(<MeasureInformation />);
    const result: HTMLElement = await findByTestId("delete-measure-button");
    fireEvent.click(result);
    const confirmDelete = await getByTestId("delete-measure-button-2");
    expect(confirmDelete).toBeInTheDocument();
    const cancelDelete = await getByTestId("cancel-delete-measure-button");
    expect(cancelDelete).toBeInTheDocument();
    fireEvent.click(confirmDelete);

    await waitFor(
      () =>
        expect(
          getByTestId("edit-measure-information-generic-error-text")
        ).toBeInTheDocument(),
      {
        timeout: 5000,
      }
    );
  });

  it("On failed delete action, user can see toast with error message", async () => {
    render(<MeasureInformation />);

    const deleteButton = await findByRole("button", { name: "Delete Measure" });
    userEvent.click(deleteButton);
    const confirmDeleteButton = await findByRole("button", {
      name: "Yes, Delete",
    });
    userEvent.click(confirmDeleteButton);
    await waitFor(
      () =>
        expect(
          getByTestId("edit-measure-information-generic-error-text")
        ).toBeInTheDocument(),
      {
        timeout: 5000,
      }
    );
    const toastErrorMessage = await findByText(
      "500: bad test oh no what happened"
    );
    expect(toastErrorMessage).toBeInTheDocument();
  });

  it("On failed update of measure, error message shown", async () => {
    render(<MeasureInformation />);
    // Click the name to trigger the inline edit
    const clickableSpan = getByTestId("inline-view-span");
    fireEvent.click(clickableSpan);
    // Have the user click enter (after input)
    useKeyPress.mockReturnValueOnce(true);
    // Type in the new value
    const input = await findByTestId("inline-edit-input");
    fireEvent.change(input, {
      target: { value: "new value" },
    });
    // Wait for rendering
    await findByTestId("inline-view-span");
    // Check expectations
    const submit = {
      id: "test measure",
      measureName: "new value",
      createdBy: "testuser@example.com", //#nosec
    };

    //const mockSetMeasure = measureContextHolder.setMeasure as jest.Mock<void>;
    expect(serviceApiMock.updateMeasure).toHaveBeenCalledWith(submit);
    const error: HTMLElement = getByTestId(
      "edit-measure-information-generic-error-text"
    );
    expect(error).toBeInTheDocument();

    //logRoles(container)
    //screen.debug()
  });

  it("Should not allow user to edit measure name if user is not the owner of measure", async () => {
    useOktaTokens.mockImplementation(() => ({
      getUserName: () => "AnotherUser",
    }));
    await render(<MeasureInformation />);

    const inlineInput = queryByTestId("inline-view-span");
    expect(inlineInput).not.toBeInTheDocument();
  });
});
