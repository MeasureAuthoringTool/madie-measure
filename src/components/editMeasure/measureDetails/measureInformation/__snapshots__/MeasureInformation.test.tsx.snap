// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeasureInformation component should render the component with a blank measure name 1`] = `
<div
  class="MeasureInformation___StyledDiv-sc-1pua5bc-2 IOWWo"
  data-testid="measure-name-edit"
>
  <div
    class="MeasureInformation__DisplayDiv-sc-1pua5bc-0 goZaEd"
  >
    <span
      class="MeasureInformation___StyledSpan2-sc-1pua5bc-5 jXgrNU"
    >
      Measure Name:
    </span>
    <div
      data-testid="outer-area"
    >
      <div
        data-testid="inline-editor"
      >
        <span
          class="InlineEdit__ClickableSpan-sc-infirb-0"
          data-testid="inline-view-span"
        >
          <svg
            aria-hidden="true"
            class="InlineEdit___StyledPencilIcon-sc-infirb-2 gvkVkb"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M13.586 3.586a2 2 0 112.828 2.828l-.793.793-2.828-2.828.793-.793zM11.379 5.793L3 14.172V17h2.828l8.38-8.379-2.83-2.828z"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
  <div
    class="MeasureInformation___StyledDiv3-sc-1pua5bc-6 ieyezC"
    data-testid="cql-library-name-display"
  >
    <span
      class="MeasureInformation___StyledSpan3-sc-1pua5bc-7 byhJob"
    >
      Measure CQL Library Name:
    </span>
    <span
      class="MeasureInformation__DisplaySpan-sc-1pua5bc-1 eEmctf"
    >
      NA
    </span>
  </div>
</div>
`;

exports[`MeasureInformation component should render the component with measure's name populated 1`] = `
<div
  class="MeasureInformation___StyledDiv-sc-1pua5bc-2 IOWWo"
  data-testid="measure-name-edit"
>
  <div
    class="MeasureInformation__DisplayDiv-sc-1pua5bc-0 goZaEd"
  >
    <span
      class="MeasureInformation___StyledSpan2-sc-1pua5bc-5 jXgrNU"
    >
      Measure Name:
    </span>
    <div
      data-testid="outer-area"
    >
      <div
        data-testid="inline-editor"
      >
        <span
          class="InlineEdit__ClickableSpan-sc-infirb-0"
          data-testid="inline-view-span"
        >
          the measure for testing
          <svg
            aria-hidden="true"
            class="InlineEdit___StyledPencilIcon-sc-infirb-2 gvkVkb"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M13.586 3.586a2 2 0 112.828 2.828l-.793.793-2.828-2.828.793-.793zM11.379 5.793L3 14.172V17h2.828l8.38-8.379-2.83-2.828z"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
  <div
    class="MeasureInformation___StyledDiv3-sc-1pua5bc-6 ieyezC"
    data-testid="cql-library-name-display"
  >
    <span
      class="MeasureInformation___StyledSpan3-sc-1pua5bc-7 byhJob"
    >
      Measure CQL Library Name:
    </span>
    <span
      class="MeasureInformation__DisplaySpan-sc-1pua5bc-1 eEmctf"
    >
      NA
    </span>
  </div>
</div>
`;
